lightning/src/ln/channelmanager.rs

```diff
@@ -3680,35 +3693,69 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> MessageSend
 				L::Target: Logger,
 {
 	fn get_and_clear_pending_msg_events(&self) -> Vec<MessageSendEvent> {
-		//TODO: This behavior should be documented. It's non-intuitive that we query
-		// ChannelMonitors when clearing other events.
-		self.process_pending_monitor_events();
+		let events = RefCell::new(Vec::new());
+		PersistenceNotifierGuard::optionally_notify(&self.total_consistency_lock, &self.persistence_notifier, || {
+			let mut result = NotifyOption::SkipPersist;
+
+			// TODO: This behavior should be documented. It's unintuitive that we query
+			// ChannelMonitors when clearing other events.
+			if self.process_pending_monitor_events() {
+				result = NotifyOption::DoPersist;
+			}
 
-		self.check_free_holding_cells();
+			self.check_free_holding_cells();
```

Comment:
I think we need to check `check_free_holding_cells` for updates to persist as well.

Reply:
Ah, sorry. Didn't look at that closely enough. Figured it was a simple merge conflict. Let me fix it.

Reply:
To confirm my understand, should we persist whenever `check_free_holding_cells` gets any failed HTLCs back from `maybe_free_holding_cell_htlcs`? Or is the condition more complicated?

https://github.com/rust-bitcoin/rust-lightning/blob/4e8bed20760ee1b53cf2fd15d4a9897780c4fab5/lightning/src/ln/channelmanager.rs#L3524

Reply:
We should also persist if we get a `ChannelMonitorUpdate` back from `maybe_free_holding_cell_htlcs` (`ChannelMonitorUpdate`s are always a surefire reason to persist - if we're out of sync when we come back up we have to force-close). Reading the code in channel that should suffice, but maybe add a comment in `maybe_free_holding_cell_htlcs` and/or `free_holding_cell_htlcs` that notes that we won't persist unless we get some values back in the return.

Reply:
Left the comment out since it's really a concern of `ChannelManager`. Made the logic in `check_free_holding_cells` explicit, though, so hopefully this is clear.